src/hotspot/share/prims/unsafe.cpp

```diff
@@ -783,21 +783,23 @@ UNSAFE_ENTRY(void, Unsafe_Park(JNIEnv *env, jobject unsafe, jboolean isAbsolute,
 
 UNSAFE_ENTRY(void, Unsafe_Unpark(JNIEnv *env, jobject unsafe, jobject jthread)) {
   if (jthread != nullptr) {
-    ThreadsListHandle tlh;
-    JavaThread* thr = nullptr;
-    oop java_thread = nullptr;
-    (void) tlh.cv_internal_thread_to_JavaThread(jthread, &thr, &java_thread);
-    if (java_thread != nullptr) {
-      // This is a valid oop.
-      if (thr != nullptr) {
-        // The JavaThread is alive.
-        Parker* p = thr->parker();
-        HOTSPOT_THREAD_UNPARK((uintptr_t) p);
-        p->unpark();
+    oop thread_oop = JNIHandles::resolve_non_null(jthread);
+    if (java_lang_Thread::thread(thread_oop) != nullptr) {
+      ThreadsListHandle tlh; // Provides memory barrier
```

Comment:
I don't understand what the comment means - what kind of memory barrier?

Reply:
Store side do:
Store to threadslist
Store to eetop

Read side do:
Read eetop
Read threadslist
Read eetop again

We want them to stay in this order.

Reply:
Perhaps the comment should change to:
```
      ThreadsListHandle tlh; // Also is a barrier between the two JavaThread* queries.
```

Reply:
I agree we need these ordering constraints, but I'm not clear what is actually guaranteeing them on either the reader or writer side. I imagine there are a number of "barriers" in relation to TLH construction, but as adding to the threadsList is done under a lock (as is setting eetop) then I'm not sure what explicit barriers exist there. Maybe `java_lang_Thread::set_thread` needs to use `release_address_field_put`?

Reply:
There are ~76 ThreadsListHandle variables in the code base. These are the only ones with comments:
```
src/hotspot/share/runtime/threadSMR.cpp:    ThreadsListHandle tlh;  // make the current ThreadsList safe for reporting
src/hotspot/share/prims/unsafe.cpp:      ThreadsListHandle tlh; // Provides memory barrier
src/hotspot/share/compiler/compileBroker.cpp:        ThreadsListHandle tlh;  // name() depends on the TLH.
src/hotspot/share/compiler/compileBroker.cpp:        ThreadsListHandle tlh;  // name() depends on the TLH.
src/hotspot/share/compiler/compileBroker.cpp:        ThreadsListHandle tlh;  // name() depends on the TLH.
src/hotspot/share/compiler/compileBroker.cpp:        ThreadsListHandle tlh;  // name() depends on the TLH.
```

The new "// Provides memory barrier" is the only one that even talks about barriers.
I'm inclined to delete the comment given that these current lines of code:
```
    if (java_lang_Thread::thread(thread_oop) != nullptr) {
      // Try to capture the live JavaThread in a ThreadsListHandle:
      ThreadsListHandle tlh;
      if (java_lang_Thread::thread(thread_oop) != nullptr) {
        // The still live JavaThread is protected by the ThreadsListHandle.
        JavaThread* thr = nullptr;
```
I think the comments added in the previous comment are enough to make this code clear.
@robehn and @dholmes-ora - are you both good with this?

Reply:
I don't mind if the "memory barrier" comment is there or not. I'm more concerned about whether the actual required ordering constraints are in fact enforced. I think `java_lang_Thread::set_thread` may need to have release semantics (and the matching `acquire` is (implicitly) within the TLH construction).

Reply:
@dholmes-ora - I'm sorry that I forgot to reply to this earlier comment in this thread:
https://github.com/openjdk/jdk/pull/13519#discussion_r1171987653

> but as adding to the threadsList is done under a lock (as is setting eetop) 

Adding a new JavaThread* to the main ThreadsList is done under a lock.
The `java_lang_Thread::set_thread()` calls are less clean:
- there's one call from`create_initial_thread()` that doesn't have the Threads_lock, but I also don't think it needs it.
- there's one call from `JavaThread::allocate_threadObj()` which happens during JNI attach that doesn't have the Threads_lock, but I don't think it needs it

- the call from `JavaThread::prepare()` has the Threads_lock
- the call from `JavaThread::start_internal_daemon()` has the Threads_lock

Of course, here's our favorite call from `ensure_join()`:
```
static void ensure_join(JavaThread* thread) {
  // We do not need to grab the Threads_lock, since we are operating on ourself.
<snip>
  OrderAccess::release();
  java_lang_Thread::set_thread(threadObj(), nullptr);
```

Obviously someone thought that call to `java_lang_Thread::set_thread()` needed a "release".

Reply:
And I need to address part of this comment:
https://github.com/openjdk/jdk/pull/13519#discussion_r1173249806

> I think java_lang_Thread::set_thread may need to have release semantics
> (and the matching acquire is (implicitly) within the TLH construction).

I agree we should put release semantics in `java_lang_Thread::set_thread()` and
I'll work on that next.

This part is puzzling me:
> the matching acquire is (implicitly) within the TLH construction

TLH construction does nothing with the `JavaThread*` stored in the `java.lang.Thread` object
unless I'm really missing something here. I would think the matching acquire would be in
any call to `java_lang_Thread::thread()`. 

So I also think that `java_lang_Thread::thread()` needs to use `address_field_acquire()`
unless I'm missing something.

Reply:
Yes you are right that makes things simpler and much neater. We end up with two acquires in the path:
```
Read eetop
Read threadslist
Read eetop again
```
but that is okay.